ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -287,12 +305,143 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
+			if err != nil && !errors.IsNotFound(err) {
+				return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
+			}
 		}
 	}
 	return ctrl.Result{}, nil
 }
 
+// CreateK8sJob creates a Kubernetes Job for the Ray Job if it doesn't exist, otherwise return the existing one.
+func (r *RayJobReconciler) CreateK8sJob(ctx context.Context, rayJobInstance *rayv1alpha1.RayJob) (string, error) {
+	var submitter_template v1.PodTemplateSpec
+	// Check if SubmitterPodTemplate is not provided
+	if rayJobInstance.Spec.SubmitterPodTemplate == nil {
+		// Set the default value
+		submitter_template = v1.PodTemplateSpec{
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{
+					{
+						Name:  "ray-job-submitter",
+						Image: "rayproject/ray:2.5.0",
+						Resources: v1.ResourceRequirements{
+							Limits: v1.ResourceList{
+								v1.ResourceCPU: resource.MustParse("1"),
+							},
+							Requests: v1.ResourceList{
+								v1.ResourceCPU: resource.MustParse("200m"),
+							},
+						},
+					},
+				},
+				RestartPolicy: v1.RestartPolicyNever,
+			},
+		}
+	} else {
+		submitter_template = *rayJobInstance.Spec.SubmitterPodTemplate.DeepCopy()
+	}
+
+	address := rayJobInstance.Status.DashboardURL
+
+	// add http:// if needed
+	if !strings.HasPrefix(address, "http://") {
+		address = "http://" + address
+	}
+
+	k8s_job_command := []string{"ray", "job", "submit", "--address", address}
+
+	// Add runtimeEnv to command
+	if len(rayJobInstance.Spec.RuntimeEnv) > 0 {
+		decodedBytes, err := base64.StdEncoding.DecodeString(rayJobInstance.Spec.RuntimeEnv)
+		if err != nil {
+			return "", fmt.Errorf("failed to decode runtimeEnv: %v: %v", rayJobInstance.Spec.RuntimeEnv, err)
+		}
+		runtimeEnv := string(decodedBytes)
+		k8s_job_command = append(k8s_job_command, "--runtime-env-json", runtimeEnv)
+	}
+
+	// Add metadata to command
+	if len(rayJobInstance.Spec.Metadata) > 0 {
+		// Check that the Ray version is at least 2.6.0.
+		// If it is, we can use the --metadata-json flag.
+		// Otherwise, we need to raise an error.
+
+		rayVersion := rayJobInstance.Spec.RayClusterSpec.RayVersion
+		constraint, _ := semver.NewConstraint(">= 2.6.0")
+		v, err := semver.NewVersion(rayVersion)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse Ray version: %v: %v", rayVersion, err)
+		}
+
+		if !constraint.Check(v) {
+			return "", fmt.Errorf("the Ray version must be at least 2.6.0 to use the metadata field")
+		}
+
+		// Convert the metadata map to a JSON string.
+		metadataBytes, err := json.Marshal(rayJobInstance.Spec.Metadata)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal metadata: %v: %v", rayJobInstance.Spec.Metadata, err)
+		}
+		metadata := string(metadataBytes)
+
+		k8s_job_command = append(k8s_job_command, "--metadata-json", metadata)
+	}
+
+	// Add jobid to command
+	if len(rayJobInstance.Status.JobId) > 0 {
+		jobid := rayJobInstance.Status.JobId
+		k8s_job_command = append(k8s_job_command, "--job-id", jobid)
```

Comment:
What's this option for?

Reply:
This passes the Ray Job ID to Ray to define the ID for the new job.  It seems currently in KubeRay we generate a Ray job ID based on the name, plus 5 random letters.  I haven't made any change to this behavior in this PR, previously `job_id` was passed via the HTTP client.  But it's possible we don't need the 5 random letters anymore and we can consider removing it in a future PR.